Roccat Kone AIMO driver

Daegalus opened the pull request:
Deciphered as much as I could figure out and updated the driver as a copy of the current Roccat driver.

Checksum passes, and the data is pulled correctly.

LED support isn't added, but I have the profile struct updated for it. Added comments to the struct for the unknown stuff and for the known stuff.

Added a gitignore for the `builddir` and `build` folders that were created and the one by .vscode, but i can remove that if you don't want it in the codebase.

Also updated the version constraint for meson_version as it was spamming me with warning about version incompatibilities with something.

stephanlachnit commented:
Thanks, I'll give it a look later! Also, please remove the Meson and gitignore changes, this is intended.

stephanlachnit commented on .gitignore:
```diff
@@ -0,0 +1,3 @@
+builddir/
+.vscode/
+build/
```
See https://github.com/libratbag/libratbag/pull/671 or https://github.com/libratbag/libratbag/pull/871.

Daegalus replied:
Removed, makes sense. I setup my globals, especially with the way Meson and Ninja work.

stephanlachnit commented on data/devices/roccat-kone-aimo-remastered.device:
```diff
@@ -0,0 +1,4 @@
+[Device]
+Name=Roccat Kone AIMO Remastered
+DeviceMatch=usb:1e7d:2e2c
+Driver=roccat
```
`Driver=roccat-kone-aimo` for now, see below.

Daegalus replied:
Done, I just copied the other one. I thought it was intended.

stephanlachnit commented on meson.build:
```diff
@@ -2,7 +2,7 @@ project('libratbag', 'c',
 	license : 'MIT/Expat',
 	default_options : [ 'c_std=gnu99', 'warning_level=2' ],
-	meson_version : '>= 0.40.0')
+	meson_version : '>= 0.50.0')
```
Ignore the warning, they are taken care of. Please revert this.

Daegalus replied:
Reverted, makes sense.

stephanlachnit commented on meson.build:
```diff
@@ -171,8 +171,9 @@ src_libratbag = [
 	'src/driver-logitech-g600.c',
-	'src/driver-roccat.c',
-	'src/driver-roccat-kone-pure.c',
+ 	'src/driver-roccat.c',
+ 	'src/driver-roccat-kone-pure.c',
```
I don't know why, but the diffs are broken here.

Daegalus replied:
There was a weird new line, I think i fixed it in the new commit I think.

stephanlachnit commented on src/driver-roccat-kone-aimo.c:
```diff
@@ -0,0 +1,876 @@
+
+struct ratbag_driver roccat_driver = {
+	.name = "Roccat Kone AIMO",
+	.id = "roccat",
```
Set this to `.id = "roccat-kone-aimo",` for now until the Roccat drivers are unified, see also https://github.com/libratbag/libratbag/pull/1001.

stephanlachnit commented on src/driver-roccat-kone-aimo.c:
```diff
@@ -0,0 +1,876 @@
+};
+
+static struct roccat_button_mapping roccat_button_mapping[] = {
+/* FIXME:	{ 0, Disabled }, */
+	{ 0, BUTTON_ACTION_NONE},
```
I think the FIXME line can be removed then

stephanlachnit commented on meson.build:
```diff
@@ -171,8 +171,9 @@ src_libratbag = [
 	'src/driver-logitech-g300.c',
 	'src/driver-logitech-g600.c',
-	'src/driver-roccat.c',
+ 	'src/driver-roccat.c',
```
weird diff

Daegalus replied:
I can't seem to fix this. I reset the file to the one in master. and then added the 1 line for the new driver. and it still has this diff.

FFY00 replied:
You have a space before the tag.

![libratbag-pr-witespace](https://user-images.githubusercontent.com/11718923/84573295-8b7b4500-ad97-11ea-803e-7dcb3e92f48c.png)

Daegalus commented:
I did with ratbagdctl.devel. I couldn't get piper.devel to work.

I'm actually going to test one more time before I'm comfortable this being merged

stephanlachnit commented:
> I did with ratbagdctl.devel. I couldn't get piper.devel to work.

What I usually do is kill my system ratbagd daemon if running, and then just run the compiled `ratbagd`. Works with Piper. But you need to place the `.device` files manually into `/usr/local/libratbag`.

Daegalus commented:
Ok, I tested it with piper, and was able to remap my scrollwheel successfully. But things I noticed:

- The UI only allows 500 or 1000 for poll rate.
- Only a small portion of the buttons are mappable from Piper. 
- Picture is obviously missing, which is fine, as I assume thats added later.
- I assume the LED stuff will show up if i implement the appropriate LED stuff?

Finally, for the space/tab diff thing. VSCode wont show it as anything other than a tab, so I am having trouble fixing it. I considered just running the `convert everything to tabs` feature in VSCode, but i found the file is a mishmash of spaces and tabs. While I honestly don't care which is used, we need to choose one and stick with it (tabs seems to be the predominant one.)

mweisshaupt1988 commented:
The code looks fine so far, you already mentioned the tab and space mix. Sadly I was not yet able to get it to run, I always get the error "ratbag error: ROCCAT ROCCAT Kone Aimo 16K: driver 'roccat-kone-aimo' does not exist". I checked the installed binary, it does contain the code. At least objdump says it does :smiley:

I'll try to get it to run and verify on my machine the next evenings.

Daegalus commented:
> The code looks fine so far, you already mentioned the tab and space mix. Sadly I was not yet able to get it to run, I always get the error "ratbag error: ROCCAT ROCCAT Kone Aimo 16K: driver 'roccat-kone-aimo' does not exist". I checked the installed binary, it does contain the code. At least objdump says it does smiley
> 
> I'll try to get it to run and verify on my machine the next evenings.

You can check if it works with `ratbagd.devel` if it doesn't might need a recompile.

mweisshaupt1988 commented:
Well, I don't know why git pull did not pull everything yesterday, but today it did. oO Nevermind, after a quick rebuild the service works now. I'm not shure what I did wrong since this was a fresh clone but it must have been something :smile: 

I could also verify that piper works with it, it does not show all buttons and while setting a macro to the back key my mouse crashed. I could reset it by unplugging and replugging it. I'll see if I can nail this down tomorrow.

This is how it looks on my machine. Obviously the SVG is not existing and the language is german :D
![Bildschirmfoto von 2020-06-16 00-34-07](https://user-images.githubusercontent.com/3426126/84712527-2119fe80-af69-11ea-9911-3bc068c44dd3.png)

But pretty good so far. Thank you :)

Daegalus commented:
Macros are definitely not something I explored too in-depth. I might need to adjust data structs or data sizes for it.

The goal for this was button mapping standard buttons and profile settings.

This will need a lot more work but it is a starting point for basic support for newer Roccat mice that Roccat tools doesn't support

mweisshaupt1988 commented:
That would be great, I miserably failed xD

Daegalus commented:
So, Ive been dealing with RL shit for a while. What are the expectations/goals of this PR? do we want to get just the mouse working first and merge that, or are we waiting on LED support from me? I know I promised some guides but have been too busy to build anything. Just wanting to know what is expected/being waited on/etc.

@stephanlachnit @FFY00

zefrof commented:
Would this driver support the Kona Aimo Remastered as well? I have a friend looking to switch and the mouse is the biggest block right now.

Daegalus commented:
> Would this driver support the Kona Aimo Remastered as well? I have a friend looking to switch (to Linux) and the mouse is the biggest block right now.

@zefrof yes, I built this driver against the Kone AIMO Remastered, as that is what I have, so it should have 100% support for the Kone AIMO Remastered.

zefrof commented:
That’s great to hear. Any ideas on when it might get merged? I completely understand if the answer is “don’t know.”

Daegalus commented:
@zefrof ya I don't know, I'm waiting on any feedback from the maintainers.

D3SOX commented:
@Daegalus Can you also add support for the Kone AIMO (non-remastered)? We tried installing your branch and changing the `DeviceMatch` to the one listed from `lsusb`
```
Bus 001 Device 006: ID 1e7d:2e27 ROCCAT Kone AIMO Mouse
```
inside `/usr/share/libratbag/roccat-kone-aimo-remastered.device`
Changing profiles works but changing the DPI doesn't do anything and gives an error when clicking on apply.

FFY00 commented:
@Daegalus sorry for the delay, I've been busy. I thought I would have an opening the past month but I didn't. Can you rebase on master? If there are any missing features, like there being no LED support, please set the number of LEDs to 0.

Daegalus commented:
@D3SOX I don't have the mouse. Unless you can provide the dump of the mouse communication for me to look through, I can't make any changes as I don't know the format.

@FFY00 will do, give me a little bit, as I switched back to windows on my primary computer. I still have a secondary that I can use to confirm things, I just need to setup the build environment on there again.

D3SOX commented:
@Daegalus If you tell me how I can do that I'll provide it

Daegalus commented:
@D3SOX you can use a windows VM with swarm installed, but you need to use wireshark and usbpcap, to monitor the traffic of the USB connection. Once you find the right USB connection. You just need to monitor the packets for the Request and Response packets over usb. the remastered ones were around 190ish kb and 130ishkb, but you will see a pattern. Usually it will be a packet that changes only a little bit with every change you make in the swarm client.

Thats the only way I know of doing it.

FFY00 commented:
No worries! I am still limited in time so I am not in a hurry :grin:

WalnussPower commented:
I have the mouse. Is there a way I can help to get this to move forward? What would I need to do? I'd really like to be able to configure my mouse from Linux.

Daegalus commented:
> I have the mouse. Is there a way I can help to get this to move forward? What would I need to do? I'd really like to be able to configure my mouse from Linux.

Sorry about that, I just got swept up in life things. I will do a rebase so that I unblock the maintainers.

WalnussPower commented:
> > I have the mouse. Is there a way I can help to get this to move forward? What would I need to do? I'd really like to be able to configure my mouse from Linux.
> 
> Sorry about that, I just got swept up in life things. I will do a rebase so that I unblock the maintainers.

No need to say "Sorry". It's volunteer work. Don't stress yourself about it!

Daegalus commented:
Looks like tests pass. the FIXMEs were there from the original Kone Pure file I used as a base. If you want I can get rid of them, but I think they are still relevant. Otherwise, rebase work is done and its ready whenever you guys are.

FFY00 commented on src/driver-roccat-kone-aimo.c:
```diff
@@ -0,0 +1,876 @@
+#define ROCCAT_NUM_DPI				5
+#define ROCCAT_LED_MAX				0 // 5 or 11 depending on how you count it.
+#define ROCCAT_MIN_DPI              100
+#define ROCCAT_MAX_DPI              16000
```
Please fix the indentation here, we use tabs (8 spaces wide), this section is using spaces.

Daegalus replied:
Done.

FFY00 commented on src/driver-roccat-kone-aimo.c:
```diff
@@ -0,0 +1,876 @@
+} __attribute__((packed));
+
+struct roccat_data {
+	uint8_t profiles[(ROCCAT_PROFILE_MAX )][ROCCAT_REPORT_SIZE_PROFILE];
+	struct roccat_settings_report settings[(ROCCAT_PROFILE_MAX )];
+	struct roccat_macro macros[(ROCCAT_PROFILE_MAX )][(ROCCAT_BUTTON_MAX)];
```
This has an extra space after `ROCCAT_PROFILE_MAX` (`(ROCCAT_PROFILE_MAX )`).

Daegalus replied:
All set

FFY00 commented on src/driver-roccat-kone-aimo.c:
```diff
@@ -0,0 +1,876 @@
+    ratbag_profile_for_each_button(profile, button)
+		roccat_read_button(button);
+
+    // No CRC check for Roccat Kone AIMO, but leaving for reference if needed for other models.
```
But this driver is specific for the AIMO :stuck_out_tongue:. Please remove the unused CRC related code, we still have it on the other driver, so we can easily grab it if needed in the future.

Daegalus replied:
Removed

FFY00 reviewed (changes requested):
I eyed the driver and it looks fine, just some minor issues. If you can confirm that all features implemented by the driver are working as expected, we can go ahead and merge this after what I pointed out is fixed.

@WalnussPower can you test the driver too? Since you have one device and are willing to help to move this forward, multiple people testing would go a long way :blush:

Daegalus commented:
Done, and I tested it and still works like it did before, but definitely wouldn't mind a sanity check.

thesecretmaster commented:
Just tried this PR with my Roccat Kone Aimo. I installed the patch, started the new ratbagd, opened piper and saw my mouse. To test it out, I tried mapping mouse right to the letter "a". When I clicked "apply" my mouse froze. When I unplugged the mouse and plugged it back in, the macro had not been applied, and opening piper causes ratbagd to segfault. Here are the relevant logs from ratbagd:

```
ratbag debug: New device: ROCCAT ROCCAT Kone Aimo 16K
ratbag debug: Using data directory '/usr/local/share/libratbag'
ratbag debug: device assigned driver roccat-kone-aimo
ratbag debug: hidraw info: bus 0x03 vendor 0x1e7d product 0x2e2c
ratbag debug: ROCCAT ROCCAT Kone Aimo 16K is device '/dev/hidraw5'.
ratbag debug: Parsing HID report descriptor
ratbag debug: Parsing HID report descriptor
ratbag debug: - HID report ID 01
ratbag debug: - HID report ID 02
ratbag debug: - HID report ID 20
ratbag debug: - HID report ID 03
ratbag debug: - HID report ID 04
ratbag debug: - HID report ID 05
ratbag debug: - HID report ID 06
ratbag debug: - HID report ID 07
ratbag debug: - HID report ID 08
ratbag debug: - HID report ID 09
ratbag debug: - HID report ID 0a
ratbag debug: - HID report ID 0c
ratbag debug: - HID report ID 0d
ratbag debug: - HID report ID 0e
ratbag debug: - HID report ID 0f
ratbag debug: - HID report ID 10
ratbag debug: - HID report ID 11
ratbag debug: - HID report ID 12
ratbag debug: - HID report ID 1a
ratbag debug: - HID report ID 1b
ratbag debug: - HID report ID 1c
ratbag raw: feature set:   04 00 80
ratbag raw: feature get:   04 01 00
ratbag raw: feature get:   06 7e 00 06 1f 01 08 00 10 00 18 00 20 00 40 00 08 00 10 00 18 00 20 00 40 00 03 00 00 08 ff 07 00 09 06 ff 1d 13 ff 00 ff 59 ff 00 00 ff fd fd 00 00 ff f4 64 00 00 ff f4 00 00 00 ff ff ff 00 00 00 ff ff 00 00 00 ff ff 50 00 00 ff ff 50 00 00 ff e6 8c 00 00 ff ff 00 00 00 ff ff 50 00 00 ff ff 50 00 00 ff e6 8c 00 00 ff e6 8c 00 00 ff e6 8c 00 00 80 01 00 00 00 00 00 00 28 29
ratbag raw: feature set:   04 00 90
ratbag raw: feature get:   04 01 00
ratbag raw: feature get:   07 4b 00 01 00 00 02 00 00 03 00 00 09 00 00 30 00 00 0d 00 00 0e 00 00 07 00 00 08 00 00 41 00 00 15 00 00 16 00 00 00 00 00 00 00 00 23 00 00 21 00 00 22 00 00 26 00 00 27 00 00 00 00 00 00 00 00 00 00 00 11 00 00 12 fd 01
ratbag raw: feature set:   04 00 00
ratbag raw: feature get:   04 03 00
ratbag raw: feature get:   04 01 00
ratbag raw: feature set:   04 00 04
ratbag raw: feature get:   04 03 00
ratbag raw: feature get:   04 01 00
ratbag raw: feature get:   08 00 ff ff ff ff ff ff ff ff ff ff ff ff ff ff ff ff ff ff ff ff ff ff ff ff ff ff ff ff ff ff ff ff ff ff ff ff ff ff ff ff ff ff ff ff ff ff ff ff ff ff ff ff ff ff ff ff ff ff ff ff ff ff ff ff ff ff ff ff ff ff ff ff ff ff ff ff ff ff ff ff ff ff ff ff ff ff ff ff ff ff ff ff ff ff ff ff ff ff ff ff ff ff ff ff ff ff ff ff ff ff ff ff ff ff ff ff ff ff ff ff ff ff ff ff ff ff ff ff ff ff ff ff ff ff ff ff ff ff ff ff ff ff ff ff ff ff ff ff ff ff ff ff ff ff ff ff ff ff ff ff ff ff ff ff ff ff ff ff ff ff ff ff ff ff ff ff ff ff ff ff ff ff ff ff ff ff ff ff ff ff ff ff ff ff ff ff ff ff ff ff ff ff ff ff ff ff ff ff ff ff ff ff ff ff ff ff ff ff ff ff ff ff ff ff ff ff ff ff ff ff ff ff ff ff ff ff ff ff ff ff ff ff ff ff ff ff ff ff ff ff ff ff ff ff ff ff ff ff ff ff ff ff ff ff ff ff ff ff ff ff ff ff ff ff ff ff ff ff ff ff ff ff ff ff ff ff ff ff ff ff ff ff ff ff ff ff ff ff ff ff ff ff ff ff ff ff ff ff ff ff ff ff ff ff ff ff ff ff ff ff ff ff ff ff ff ff ff ff ff ff ff ff ff ff ff ff ff ff ff ff ff ff ff ff ff ff ff ff ff ff ff ff ff ff ff ff ff ff ff ff ff ff ff ff ff ff ff ff ff ff ff ff ff ff ff ff ff ff ff ff ff ff ff ff ff ff ff ff ff ff ff ff ff ff ff ff ff ff ff ff ff ff ff ff ff ff ff ff ff ff ff ff ff ff ff ff ff ff ff ff ff ff ff ff ff ff ff ff ff ff ff ff ff ff ff ff ff ff ff ff ff ff ff ff ff ff ff ff ff ff ff ff ff ff ff ff ff ff ff ff ff ff ff ff ff ff ff ff ff ff ff ff ff ff ff ff ff ff ff ff ff ff ff ff ff ff ff ff ff ff ff ff ff ff ff ff ff ff ff ff ff ff ff ff ff ff ff ff ff ff ff ff ff ff ff ff ff ff ff ff ff ff ff ff ff ff ff ff ff ff ff ff ff ff ff ff ff ff ff ff ff ff ff ff ff ff ff ff ff ff ff ff ff ff ff ff ff ff ff ff ff ff ff ff ff ff ff ff ff ff ff ff ff ff ff ff ff ff ff ff ff ff ff ff ff ff ff ff ff ff ff ff ff ff ff ff ff ff ff ff ff ff ff ff ff ff ff ff ff ff ff ff ff ff ff ff ff ff ff ff ff ff ff ff ff ff ff ff ff ff ff ff ff ff ff ff ff ff ff ff ff ff ff ff ff ff ff ff ff ff ff ff ff ff ff ff ff ff ff ff ff ff ff ff ff ff ff ff ff ff ff ff ff ff ff ff ff ff ff ff ff ff ff ff ff ff ff ff ff ff ff ff ff ff ff ff ff ff ff ff ff ff ff ff ff ff ff ff ff ff ff ff ff ff ff ff ff ff ff ff ff ff ff ff ff ff ff ff ff ff ff ff ff ff ff ff ff ff ff ff ff ff ff ff ff ff ff ff ff ff ff ff ff ff ff ff ff ff ff ff ff ff ff ff ff ff ff ff ff ff ff ff ff ff ff ff ff ff ff ff ff ff ff ff ff ff ff ff ff ff ff ff ff ff ff ff ff ff ff ff ff ff ff ff ff ff ff ff ff ff ff ff ff ff ff ff ff ff ff ff ff ff ff ff ff ff ff ff ff ff ff ff ff ff ff ff ff ff ff ff ff ff ff ff ff ff ff ff ff ff ff ff ff ff ff ff ff ff ff ff ff ff ff ff ff ff ff ff ff ff ff ff ff ff ff ff ff ff ff ff ff ff ff ff ff ff ff ff ff ff ff ff ff ff ff ff ff ff ff ff ff ff ff ff ff ff ff ff ff ff ff ff ff ff ff ff ff ff ff ff ff ff ff ff ff ff ff ff ff ff ff ff ff ff ff ff ff ff ff ff ff ff ff ff ff ff ff ff ff ff ff ff ff ff ff ff ff ff ff ff ff ff ff ff ff ff ff ff ff ff ff ff ff ff ff ff ff ff ff ff ff ff ff ff ff ff ff ff ff ff ff ff ff ff ff ff ff ff ff ff ff ff ff ff ff ff ff ff ff ff ff ff ff ff ff ff ff ff ff ff ff ff ff ff 98 3b b6 34 85 a3 3a fe bd 52 83 75 11 78 f7 96 e1 e1 00 c0 03 40 40 00 00 00 00 80 04 40 6b 04 ff 1f 00 00 00 00 01 06 ff 1f 2b 65 46 01 01 00 00 00 00 00 00 00 02 00 00 d8 03 00 ef 00 00 ab 56 34 04 01 00 ce 00 00 ff 00 00 00 00 00 ff ff c6 c4 af 87 97 74 b6 fe 75 67 5f 81 b3 14 07 71 ca ed 67 75 3d 55 31 f6 5c 71 60 8f 25 b4 ff ff ff ff e0 18 00 10 b5 22 00 00 c7 22 00 00 bd 22 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 bf 22 00 00 00 00 00 00 00 00 00 00 c1 22 00 00 4b 32 00 00 85 46 00 00 b5 46 00 00 e7 46 00 00 19 47 00 00 4b 47 00 00 7d 47 00 00 af 47 00 00 e1 47 00 00 13 48 00 00 3d 48 00 00 c5 22 00 00 c5 22 00 00 c5 22 00 00 c5 22 00 00 9b 68 00 00 a5 73 00 00 a9 44 00 00 bf 44 00 00 a1 43 00 00 09 44 00 00 6d 68 00 00 c7 22 00 00 e3 29 00 00 c7 22 00 00 c7 22 00 00 c7 22 00 00 c7 22 00 00 c7 22 00 00 c5 22 00 00 c5 22 00 00 b9 2a 00 00 c5 22 00 00 03 48 85 46 07 f0 be f8 00 48 00 47 27 40 00 00 e0 18 00 10 70 b5 61 49 00 20 49 6a c9 05 09 0f 10 29 29 d2 0b 00 7b 44 1b 79 db 18 9f 44 24 07 09 0b 0d 0f 11 13 15 17 19 1b 1d 1f 21 23 58 48 1a e0 58 48 18 e0 58 48 16 e0 58 48 14 e0 58 48 12 e0 58 48 10 e0 58 48 0e e0 58 48 0c e0 58 48 0a e0 58 48 08 e0 58 48 06 e0 58 48 04 e0 58 48 02 e0 58 48 00 e0 58 48 48 4c 61 6a c9 06 89 0e 89 1c 06 f0 63 ff 55 4a 11 6b 55 4d 89 07 89 0f 11 d0 01 29 04 d0 02 29 2b d0 03 29 2a d1 0d e0 10 68 80 07 80 0f 06 d0 01 28 04 d0 02 28 27 d0 03 28 1f d1 24 e0 4b 48 28 60 1b e0 10 68 80 07 80 0f 06 d0 01 28 0a d0 02 28 19 d0 03 28 11 d1 16 e0 a0 68 c0 05 80 0f 05 d0 42 48 09 e0 a0 68 c0 05 80 0f f9 d1 a0 68 3e 49 c0 06 c0 0e 40 1c 48 43 28 60 91 6b 28 68 06 f0 28 ff 28 60 70 bd 00 20 f6 e7 f8 b4 38 4a 90 6b 20 21 88 43 90 63 22 4d 00 26 2e 62 00 96 00 bf 00 98 40 1c 00 90 c8 28 f9 d3 2d 48 01 21 01 60 41 60 46 60 41 60 43 68 db 07 fc d0 23 27 af 60 93 6b 80 24 a3 43 00 6c dc 02 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 f5 00 01 00 00 00 03 00 01 01 00 00 00 00 00 05 04 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 01 00 00 00 00 00 00 00 00 00 ff 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 ff 00 00 00 00 00 00 00 00 00 00 00 00 00 00 08 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 ff 07 00 00 34 00 00 00 00 00 00 00 00 00 00 00 00 00 01 00 00 00 00 01 02 e7 ff ff ff ff 0f 00 00 00 00 00 00 00 00 00 00 00 00 00 00 ff 0f 06 00 81 d0 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 ff 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 37 85 00 00 00 04 00 00 00 bc 00 00 00 c0 00 00 0c 00 00 00 16 00 00 00 00 00 00 00 0f 00 00 00 00 00 00 00 0f 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 12 00 00 00 01 00 00 00 0e 00 14 00 00 00 00 00 00 00 00 00 00 02 00 00 00 00 00 00 00 00 00 00 01 00 00 00 00 00 00 00 00 00 00 00 00 00 00 41 00 20 40 00 00 00 40 41 00 20 08 00 00 00 80 41 00 20 40 00 00 00 c0 41 00 20 40 00 00 00 00 42 00 20 40 00 00 00 40 42 00 20 40 00 00 00 80 42 00 20 40 00 00 00 c0 42 00 20 40 00 00 00 00 43 00 20 40 00
ratbag raw: checksum computed: 0xccf4, checksum given: 0x0040
ratbag error: wrong checksum while reading the macro of button 4 of profile 0.
ratbag raw: profile: 0 ../src/driver-roccat-kone-aimo.c:791
ratbag raw: feature set:   04 01 80
ratbag raw: feature get:   04 01 00
ratbag raw: feature get:   06 7e 01 06 1f 01 08 00 10 00 18 00 20 00 40 00 08 00 10 00 18 00 20 00 40 00 03 00 00 08 ff 07 00 09 06 ff 1d 13 ff 00 ff 59 ff 00 00 ff fd fd 00 00 ff f4 64 00 00 ff f4 00 00 00 ff ff 00 ff 00 00 ff 00 ff 00 00 ff 00 ff 32 00 ff 00 ff 32 00 ff 00 ff 64 00 ff 00 ff 00 00 ff 00 ff 32 00 ff 00 ff 32 00 ff 00 ff 64 00 ff 00 ff 64 00 ff 00 ff 64 00 86 01 00 00 00 00 00 00 7b 28
ratbag raw: feature set:   04 01 90
ratbag raw: feature get:   04 01 00
ratbag raw: feature get:   07 4b 01 01 00 00 02 00 00 03 00 00 09 00 00 0a 00 00 0d 00 00 0e 00 00 07 00 00 08 00 00 41 00 00 15 00 00 16 00 00 00 00 00 00 00 00 23 00 00 21 00 00 22 00 00 26 00 00 27 00 00 00 00 00 00 00 00 00 00 00 11 00 00 12 00 00
ratbag raw: profile: 1 ../src/driver-roccat-kone-aimo.c:791
ratbag raw: feature set:   04 02 80
ratbag raw: feature get:   04 01 00
ratbag raw: feature get:   06 7e 02 06 1f 01 08 00 10 00 18 00 20 00 40 00 08 00 10 00 18 00 20 00 40 00 03 00 00 08 ff 07 00 09 06 ff 1d 13 ff 00 ff 59 ff 00 00 ff fd fd 00 00 ff f4 64 00 00 ff f4 00 00 00 ff ff 00 00 ff 00 ff 00 00 ff 00 ff 00 7f 98 00 ff 00 7f 98 00 ff 00 ff 32 00 ff 00 00 ff 00 ff 00 7f 98 00 ff 00 7f 98 00 ff 00 ff 32 00 ff 00 ff 32 00 ff 00 ff 32 00 88 01 00 00 00 00 00 00 4e 27
ratbag raw: feature set:   04 02 90
ratbag raw: feature get:   04 01 00
ratbag raw: feature get:   07 4b 02 01 00 00 02 00 00 03 00 00 09 00 00 0a 00 00 0d 00 00 0e 00 00 07 00 00 08 00 00 41 00 00 15 00 00 16 00 00 00 00 00 00 00 00 23 00 00 21 00 00 22 00 00 26 00 00 27 00 00 00 00 00 00 00 00 00 00 00 11 00 00 12 00 00
ratbag raw: profile: 2 ../src/driver-roccat-kone-aimo.c:791
ratbag raw: feature set:   04 03 80
ratbag raw: feature get:   04 01 00
ratbag raw: feature get:   06 7e 03 06 1f 01 08 00 10 00 18 00 20 00 40 00 08 00 10 00 18 00 20 00 40 00 03 00 00 08 ff 07 00 09 06 ff 1d 13 ff 00 ff 59 ff 00 00 ff fd fd 00 00 ff f4 64 00 00 ff f4 00 00 00 ff ff c8 ff 00 00 ff c8 ff 00 00 ff 64 ff 05 00 ff 64 ff 05 00 ff 00 ff 0a 00 ff c8 ff 00 00 ff 64 ff 05 00 ff 64 ff 05 00 ff 00 ff 0a 00 ff 00 ff 0a 00 ff 00 ff 0a 00 85 01 00 00 00 00 00 00 48 2a
ratbag raw: feature set:   04 03 90
ratbag raw: feature get:   04 01 00
ratbag raw: feature get:   07 4b 03 01 00 00 02 00 00 03 00 00 09 00 00 0a 00 00 0d 00 00 0e 00 00 07 00 00 08 00 00 41 00 00 15 00 00 16 00 00 00 00 00 00 00 00 23 00 00 21 00 00 22 00 00 26 00 00 27 00 00 00 00 00 00 00 00 00 00 00 11 00 00 12 00 00
ratbag raw: profile: 3 ../src/driver-roccat-kone-aimo.c:791
ratbag raw: feature set:   04 04 80
ratbag raw: feature get:   04 01 00
ratbag raw: feature get:   06 7e 04 06 1f 01 08 00 10 00 18 00 20 00 40 00 08 00 10 00 18 00 20 00 40 00 03 00 00 08 ff 07 00 09 06 ff 1d 13 ff 00 ff 59 ff 00 00 ff fd fd 00 00 ff f4 64 00 00 ff f4 00 00 00 ff ff 00 00 ff 00 ff 00 00 ff 00 ff 4b 4b ff 00 ff 4b 4b ff 00 ff 96 96 ff 00 ff 00 00 ff 00 ff 4b 4b ff 00 ff 4b 4b ff 00 ff 96 96 ff 00 ff 96 96 ff 00 ff 96 96 ff 00 81 01 00 00 00 00 00 00 29 2d
ratbag raw: feature set:   04 04 90
ratbag raw: feature get:   04 01 00
ratbag raw: feature get:   07 4b 04 01 00 00 02 00 00 03 00 00 09 00 00 0a 00 00 0d 00 00 0e 00 00 07 00 00 08 00 00 41 00 00 15 00 00 16 00 00 00 00 00 00 00 00 23 00 00 21 00 00 22 00 00 26 00 00 27 00 00 00 00 00 00 00 00 00 00 00 11 00 00 12 00 00
ratbag raw: profile: 4 ../src/driver-roccat-kone-aimo.c:791
ratbag raw: feature get:   05 04 00
ratbag raw: 'ROCCAT ROCCAT Kone Aimo 16K' is in profile 0
ratbag debug: driver match found: Roccat Kone AIMO
hidraw5: "ROCCAT ROCCAT Kone Aimo 16K", 5 profiles
ratbag debug: New device: ROCCAT ROCCAT Kone Aimo 16K
ratbag debug: Using data directory '/usr/local/share/libratbag'
ratbag debug: device assigned driver roccat-kone-aimo
ratbag debug: hidraw info: bus 0x03 vendor 0x1e7d product 0x2e2c
ratbag debug: ROCCAT ROCCAT Kone Aimo 16K is device '/dev/hidraw6'.
ratbag debug: Parsing HID report descriptor
ratbag debug: Parsing HID report descriptor
```

FFY00 commented:
Misclick, sorry!

FFY00 commented:
Can you do a debug build and provide the backtrace?

thesecretmaster commented:
Yes, absolutely, but I'm pretty new to ratbagd, could you tell me how / link to instructions?

FFY00 commented:
We use Meson, so it should be pretty simple, just pass `--buildtype debug` (the command should become something like `meson --buildtype debug builddir`) and then you can run with gdb and get the backtrace as you would any other project.

I am on my phone, so I can't provide a demo. Let me know if you have any trouble.

thesecretmaster commented:
Here's the trace:

```
#0  ratbag_button_get_macro (button=0x5555555c3ae0) at ../src/libratbag.c:1687
#1  0x000055555555da77 in ratbagd_button_get_macro
    (bus=0x5555555bf3d0, path=0x5555555d2618 "/org/freedesktop/ratbag1/button/hidraw5/p0/b4", interface=0x5555555cc210 "org.freedesktop.ratbag1.Button", property=0x555555593834 "Mapping", reply=0x5555555e8820, userdata=0x5555555c3c10, error=0x7fffffffe1d0) at ../ratbagd/ratbagd-button.c:171
#2  0x000055555555e11d in ratbagd_button_get_mapping
    (bus=0x5555555bf3d0, path=0x5555555d2618 "/org/freedesktop/ratbag1/button/hidraw5/p0/b4", interface=0x5555555cc210 "org.freedesktop.ratbag1.Button", property=0x555555593834 "Mapping", reply=0x5555555e8820, userdata=0x5555555c3c10, error=0x7fffffffe1d0) at ../ratbagd/ratbagd-button.c:284
#3  0x00007ffff7f1f75b in  () at /lib/x86_64-linux-gnu/libsystemd.so.0
#4  0x00007ffff7f20f4a in  () at /lib/x86_64-linux-gnu/libsystemd.so.0
#5  0x00007ffff7f2202e in  () at /lib/x86_64-linux-gnu/libsystemd.so.0
#6  0x00007ffff7f2473d in  () at /lib/x86_64-linux-gnu/libsystemd.so.0
#7  0x00007ffff7f0b158 in  () at /lib/x86_64-linux-gnu/libsystemd.so.0
#8  0x00007ffff7f0b597 in  () at /lib/x86_64-linux-gnu/libsystemd.so.0
#9  0x00007ffff7ef2707 in  () at /lib/x86_64-linux-gnu/libsystemd.so.0
#10 0x00007ffff7ef31a1 in sd_event_dispatch ()
    at /lib/x86_64-linux-gnu/libsystemd.so.0
#11 0x00007ffff7ef6bd8 in sd_event_run ()
    at /lib/x86_64-linux-gnu/libsystemd.so.0
#12 0x00007ffff7ef6dff in sd_event_loop ()
    at /lib/x86_64-linux-gnu/libsystemd.so.0
#13 0x000055555555c56d in ratbagd_run (ctx=0x5555555bee40)
    at ../ratbagd/ratbagd.c:527
#14 0x000055555555c70b in main (argc=1, argv=0x7fffffffe628)
    at ../ratbagd/ratbagd.c:652
```

Daegalus commented:
You know, now that I think about it. I never made sure macros worked at all. So the macro data structure might not work at all. Mostly since I don't use macros. I don't know when I will get to it. I need to setup all my wireshark stuff too to figure out the structure of their macros. If anyone gets to it sooner and provides me with an updated data structure. I will be more than happy to update it in the code before I get around to doing the digging myself.

FFY00 commented:
If you don't have time right now, we can remove the macro code and add it later.

thesecretmaster commented:
I'm happy to test or pull data or whatever's needed, but I'll just need things explained in reasonably simple terms to me because I'm not very familiar with USB debugging or ratbag.

Daegalus commented:
> If you don't have time right now, we can remove the macro code and add it later.

@FFY00 Whats the recommended way to disable it? Just remove the functions? Or do I flip a flag somewhere? or do i just no-op the functions?

@thesecretmaster Its not hard just takes some setup. You need to have windows or a windows VM to work with. In Windows you install Wireshark and USBpcap. Once you do that, you install Roccat software, load up wireshark, find the mouse and then just look at packets. I usually change a setting in roccat, save it, and see what packets are sent. The tricky part is there is constant spam from the mouse for movement/laser positioning stuff. Its easy to filter if you set a query for a size minimum of the packet. Then you just need to figure out the structure and what does what in the packet. Not hard but very time consuming.

Daegalus commented:
Ok, in the spirit of hacktoberfest, I am going to dedicate some quality time on this. So I am currently re-installing Linux so I can properly test the mouse. I am also going to try and deduce the proper format for macros. worst case I dont get it working, but I will put in some time to figure it out.

perigoso commented:
@Daegalus do you have time to finish this up? looks like It's in the final stretch, then we can finally merge this

Daegalus commented:
> @Daegalus do you have time to finish this up? looks like It's in the final stretch, then we can finally merge this

I will, just got busy again. To be transparent, I'm becoming a father in January, and all the prep, family, etc to get ready is taking up time in addition to my day job. I will get back to this soon.

perigoso commented:
Ah don't worry too much about it, family comes first.

Sorry for being pushy, just trying to make surre things dont get forgotten :sweat_smile:

WalnussPower commented:
Is there a way I can help this move forward?

Daegalus commented:
Things are getting back to normalish so I might take one last crack at it, but if I can't get macros working, I will disable macros and let it merge. Then someone else with more C/ratbag experience can finish it up. I have the data structure, I just am not quite sure how to get it to read/send 2 packets for 1 data structure. Mostly because I haven't internalized the ratbag design to figure it out.

staticssleever668 commented:
Hello, has there been any progress on this?

By the way, if you re-base this on master this will still require some changes after the merge of #1277.
Here are they so that you can focus on more important code instead:
```diff
diff --git a/src/driver-roccat-kone-aimo.c b/src/driver-roccat-kone-aimo.c
index a5813ae3..e47b9ef3 100644
--- a/src/driver-roccat-kone-aimo.c
+++ b/src/driver-roccat-kone-aimo.c
@@ -739,9 +739,10 @@ roccat_write_macro(struct ratbag_button *button,
 }
 
 static int
-roccat_write_button(struct ratbag_button *button,
-		const struct ratbag_button_action *action)
+roccat_write_button(struct ratbag_button *button)
 {
+	const struct ratbag_button_action *action = &button->action;
+
 	struct ratbag_profile *profile = button->profile;
 	struct ratbag_device *device = profile->device;
 	struct roccat_data *drv_data = ratbag_get_drv_data(device);
@@ -775,8 +776,7 @@ roccat_write_button(struct ratbag_button *button,
 }
 
 static int
-roccat_write_resolution_dpi(struct ratbag_resolution *resolution,
-			int dpi_x, int dpi_y)
+roccat_write_resolution_dpi(struct ratbag_resolution *resolution)
 {
 	struct ratbag_profile *profile = resolution->profile;
 	struct ratbag_device *device = profile->device;
@@ -785,6 +785,9 @@ roccat_write_resolution_dpi(struct ratbag_resolution *resolution,
 	uint8_t *buf;
 	int rc;
 
+	const unsigned int dpi_x = resolution->dpi_x;
+	const unsigned int dpi_y = resolution->dpi_y;
+
 	if (dpi_x < ROCCAT_MIN_DPI || dpi_x > ROCCAT_MAX_DPI || dpi_x % 50)
 		return -EINVAL;
 	if (dpi_y < ROCCAT_MIN_DPI || dpi_y > ROCCAT_MAX_DPI || dpi_y % 50)
@@ -960,13 +963,50 @@ roccat_remove(struct ratbag_device *device)
 	free(ratbag_get_drv_data(device));
 }
 
+static int
+roccat_commit(struct ratbag_device *device)
+{
+
+	int rc = 0;
+	struct ratbag_button *button = NULL;
+	struct ratbag_profile *profile = NULL;
+	struct ratbag_resolution *resolution = NULL;
+
+	ratbag_device_for_each_profile(device, profile) {
+		if (!profile->dirty)
+			continue;
+
+		rc = roccat_write_profile(profile);
+		if (rc)
+			return rc;
+
+		ratbag_profile_for_each_resolution(profile, resolution) {
+			if (!resolution->dirty)
+				continue;
+
+			rc = roccat_write_resolution_dpi(resolution);
+			if (rc)
+				return rc;
+		}
+
+		ratbag_profile_for_each_button(profile, button) {
+			if (!button->dirty)
+				continue;
+
+			rc = roccat_write_button(button);
+			if (rc)
+				return rc;
+		}
+	}
+
+	return 0;
+}
+
 struct ratbag_driver roccat_kone_aimo_driver = {
 	.name = "Roccat Kone AIMO",
 	.id = "roccat-kone-aimo",
 	.probe = roccat_probe,
 	.remove = roccat_remove,
-	.write_profile = roccat_write_profile,
+	.commit = roccat_commit,
 	.set_active_profile = roccat_set_current_profile,
-	.write_button = roccat_write_button,
-	.write_resolution_dpi = roccat_write_resolution_dpi,
 };
```

Daegalus commented:
@staticssleever668 Thank you for the patch, definitely saved me a lot of time getting the branch updated.

I am not sure who the current active maintainers are, but I am throwing in the towel on macros. I have been away from the code for too long, and I am just not good enough with C.

So, if I can just be advised on how we can proceed to getting this PR merged so someone more experienced with C and/or Ratbag can fill in the last little bit. All the data structures are accurate and commented to the best of my knowledge. I just couldn't figure out how to process 2 separate USB packets/reports as 1 for macros with what I knew. I can also be reached out to whoever picks this up for answering questions.

If you would like me to disable Macros, I can add a DISABLE flag, and leave the current code in, but short circuit it on the first line of the function returning `-ENOSUP` but leave my code as reference.

Anyways just let me know, and I will make the few changes needed to merge this in without macros working or with macros disabled.

staticssleever668 commented on src/driver-roccat-kone-aimo.c:
```diff
@@ -0,0 +1,1012 @@
+	ratbag_profile_for_each_resolution(profile, resolution) {
+		dpi_x = setting_report->xres[resolution->index] * 50;
+		dpi_y = setting_report->yres[resolution->index] * 50;
+		//if (!(setting_report->dpi_mask & (1 << resolution->index))) {
+		//	/* the profile is disabled, overwrite it */
+		//	dpi_x = 0;
+		//	dpi_y = 0;
+		//}
```
Any reason this is commented out?

staticssleever668 commented on src/driver-roccat-kone-aimo.c:
```diff
@@ -0,0 +1,1012 @@
+	uint16_t checksum;
+} __attribute__((packed));
+
+typedef struct {
+	uint8_t keycode;
+	uint8_t flag;
+	uint16_t time;
+} roccat_macro_keys;
```
Please, revert the change of this to a `typedef`. There is nothing bad with this, it's just that it's not used much in libratbag's code base.

staticssleever668 commented:
By the way, conversations opened by amfern still hold true, please, address them. If there is anything you don't understand there, I will help you out. :)


> I just couldn't figure out how to process 2 separate USB packets/reports as 1 for macros with what I knew.

Start out by making `macroP1`, `macroP2` and `macroC` on the stack, here are some changes I would do first:
```diff
diff --git a/src/driver-roccat-kone-aimo.c b/src/driver-roccat-kone-aimo.c
index 97d2fdc8..39aa933d 100644
--- a/src/driver-roccat-kone-aimo.c
+++ b/src/driver-roccat-kone-aimo.c
@@ -146,8 +146,8 @@ struct roccat_macro_page2 {
 } __attribute__((packed));
 
 struct roccat_macro_combined {
-	struct roccat_macro_page1 *page1;
-	struct roccat_macro_page2 *page2;
+	struct roccat_macro_page1 page1;
+	struct roccat_macro_page2 page2;
 } __attribute__((packed));
 
 enum roccat_macro_action_type {
@@ -591,31 +591,23 @@ roccat_write_macro(struct ratbag_button *button,
 {
 	struct ratbag_device *device;
 	struct roccat_macro *macro;
-	struct roccat_macro_page1 *macroP1 = malloc(sizeof(struct roccat_macro_page1));
-	struct roccat_macro_page2 *macroP2 = malloc(sizeof(struct roccat_macro_page2));
-	struct roccat_macro_combined *macroC = malloc(sizeof(struct roccat_macro_combined));
+	// TODO: rename this to `macro_combined` or whatever fits.
+	struct roccat_macro_combined macroC;
 	struct roccat_data *drv_data;
-	uint8_t *buf1;
-	uint8_t *buf2;
-	uint8_t *buf;
 	unsigned count = 0;
 	int rc;
 
+	// TODO: check of both of these actually make the code clearer,
+	// otherwise simply use `macroC.page1` and `macroC.page2` directly.
+	struct roccat_macro_page1* macroP1 = &macroC.page1;
+	struct roccat_macro_page2* macroP2 = &macroC.page2;
+
 	if (action->type != RATBAG_BUTTON_ACTION_TYPE_MACRO)
 		return 0;
 
 	device = button->profile->device;
 	drv_data = ratbag_get_drv_data(device);
 	macro = &drv_data->macros[button->profile->index][button->index];
-	macroC->page1 = macroP1;
-	macroC->page2 = macroP2;
-	buf1 = (uint8_t*)macroP1;
-	buf2 = (uint8_t*)macroP2;
-	buf = (uint8_t*)macroC;
-
-	memset(buf1, 0, ROCCAT_REPORT_SIZE_MACRO);
-	memset(buf2, 0, ROCCAT_REPORT_SIZE_MACRO);
-	memset(buf, 0, ROCCAT_REPORT_SIZE_MACRO * 2);
 
 	for (unsigned int i = 0; i < MAX_MACRO_EVENTS && count < ROCCAT_MAX_MACRO_LENGTH; i++) {
 		if (action->macro->events[i].type == RATBAG_MACRO_EVENT_INVALID)
@@ -675,22 +667,16 @@ roccat_write_macro(struct ratbag_button *button,
 	macroP2->profile = button->profile->index;
 	macroP2->page = 0x02;
 
-	macroP2->checksum = roccat_compute_crc(buf, ROCCAT_REPORT_SIZE_MACRO * 2);
+	macroP2->checksum = roccat_compute_crc((uint8_t*) &macroC, sizeof(macroC));
 	macroP2->terminator = 0x4A;
 
 	rc = ratbag_hidraw_set_feature_report(device, ROCCAT_REPORT_ID_MACRO,
-				buf1, ROCCAT_REPORT_SIZE_MACRO);
+				(uint8_t*)macroP1, ROCCAT_REPORT_SIZE_MACRO);
 	if (rc < 0) {
-		free(macroP1);
-		free(macroP2);
-		free(macroC);
 		return rc;
 	}
 
 	if (rc != ROCCAT_REPORT_SIZE_MACRO) {
-		free(macroP1);
-		free(macroP2);
-		free(macroC);
 		return -EIO;
 	}
 
@@ -699,25 +685,16 @@ roccat_write_macro(struct ratbag_button *button,
 		log_error(device->ratbag,
 			"Error while waiting for the device to be ready: %s (%d)\n",
 			strerror(-rc), rc);
-		free(macroP1);
-		free(macroP2);
-		free(macroC);
 		return rc;
 	}
 
 	rc = ratbag_hidraw_set_feature_report(device, ROCCAT_REPORT_ID_MACRO,
-				buf2, ROCCAT_REPORT_SIZE_MACRO);
+				(uint8_t*)macroP2, ROCCAT_REPORT_SIZE_MACRO);
 	if (rc < 0) {
-		free(macroP1);
-		free(macroP2);
-		free(macroC);
 		return rc;
 	}
 
 	if (rc != ROCCAT_REPORT_SIZE_MACRO) {
-		free(macroP1);
-		free(macroP2);
-		free(macroC);
 		return -EIO;
 	}
 
@@ -726,15 +703,9 @@ roccat_write_macro(struct ratbag_button *button,
 		log_error(device->ratbag,
 			"Error while waiting for the device to be ready: %s (%d)\n",
 			strerror(-rc), rc);
-		free(macroP1);
-		free(macroP2);
-		free(macroC);
 		return rc;
 	}
 
-	free(macroP1);
-	free(macroP2);
-	free(macroC);
 	return rc;
 }
```

Daegalus commented:
> The `Add assert to make sure this struct size actually equals the 1026` comments by amfern are actually very important - do these first as right now the size of is 900-something, in other word it's _different_. Because of this I feel like the code in it's current state doesn't work because at the very least the checksum calculation is going to be wrong.

I believe I took care of that. added the asserts and unions. Fixed up some sizing after a secondary look at the USB packets.

> For me it looks like you are already at the finish line, so let's just make a final push.

Ok, I will push through. I just committed what I believe should work. I am currently trying to setup my ratbagd and piper to see if it actually works.

My concerns currently are in the `roccat_read_button` function. I am not sure how it's getting 2 packets worth of data into 1 struct, unless `drv_data` has magic to take care of that.

Also `macro.msg.page1.msg` to access fields in each page seems....messy. Any thoughts? I could set the pointer to page msg to a variable and use that instead.

staticssleever668 commented:
> My concerns currently are in the `roccat_read_button` function. I am not sure how it's getting 2 packets worth of data into 1 struct, unless `drv_data` has magic to take care of that.

There is no magic at all, the data is received *a single* time on line 542 by a `ratbag_hidraw_get_feature_report` call.

https://github.com/libratbag/libratbag/blob/26766fcd944ab9119a830be3dcb3406722dcddfc/src/driver-roccat-kone-aimo.c#L542

If I understood you right, official software receives *two* packets separately, if so, I imagine is should be something like this: 

```diff

diff --git a/src/driver-roccat-kone-aimo.c b/src/driver-roccat-kone-aimo.c
index d6480133..4b2bae89 100644
--- a/src/driver-roccat-kone-aimo.c
+++ b/src/driver-roccat-kone-aimo.c
@@ -540,7 +540,16 @@ roccat_read_button(struct ratbag_button *button)
 		buf[0] = ROCCAT_REPORT_ID_MACRO;
 		buf[ROCCAT_REPORT_SIZE_MACRO] = ROCCAT_REPORT_ID_MACRO;
 		rc = ratbag_hidraw_get_feature_report(device, ROCCAT_REPORT_ID_MACRO,
-					buf, ROCCAT_REPORT_SIZE_MACRO);
+					(uint8_t*)macro->msg.page1.data, ROCCAT_REPORT_SIZE_MACRO);
+		if (rc != ROCCAT_REPORT_SIZE_MACRO) {
+			log_error(device->ratbag,
+				"Unable to retrieve the macro for button %d of profile %d: %s (%d)\n",
+				button->index, button->profile->index,
+				rc < 0 ? strerror(-rc) : "not read enough", rc);
+			// FIXME: return an error here or handle this somehow.
+		}
+		rc = ratbag_hidraw_get_feature_report(device, ROCCAT_REPORT_ID_MACRO,
+					(uint8_t*)macro->msg.page2.data, ROCCAT_REPORT_SIZE_MACRO);
 		if (rc != ROCCAT_REPORT_SIZE_MACRO) {
 			log_error(device->ratbag,
 				"Unable to retrieve the macro for button %d of profile %d: %s (%d)\n",
```

---

> Also `macro.msg.page1.msg` to access fields in each page seems....messy. Any thoughts? I could set the pointer to page msg to a variable and use that instead.

I took a quick look, it's indeed quite messy... Your idea sounds good, but I'll take a more in-depth look once I have a bit more time.

Daegalus commented:
So I was making the changes, and I noticed a Roccat device was added a year ago, the Kone EMP. I was looking through it's code, and it has a lot of the same value and settings, it also handles the 2 packet macros. And while the LEDs don't match, and a lot of the extra settings. Those just rely on me updating the structs and LED code.

I am going to do a local test on maybe reworking the Kone EMP code to work with the data from the Kone AIMO. Since the Kone EMP was already merged, I already know it passed the review bar.

Daegalus commented:
Ok, I ported it to the new code. I made the appropriate changes. Profile, Settings, and LEDs work fine.

For some reason though, all the buttons show up as Unknown and I have yet to figure out why. So I can't check it macros work until buttons are parsed properly.

Daegalus commented:
Ok, did a small amount of log debugging, and Macros and Buttons are being read correctly, and based on the logging entries i made, they are correctly being converted to and from ratbag actions. Not sure why `ratbagctl` is returning `UNKNOWN` for all for all of them.

staticssleever668 commented:
That's weird to say the least. How do you know they are read correctly? 



> Not sure why `ratbagctl` is returning `UNKNOWN` for all for all of them.

It does so

staticssleever668 commented on src/driver-roccat-kone-aimo.c:
```diff
@@ -0,0 +1,1220 @@
+	struct roccat_macro *macro;
+	int rc;
+
+	action = roccat_button_to_action(button->profile, button->index);
```
I think this may have been intended to be:
```suggestion
	button->action = roccat_button_to_action(button->profile, button->index);
```

I think this is the reason libratbag says "UNKNOWN" on buttons.

Daegalus replied:
Maybe, the next lines do that. it calls `ratbag_button_set_action` which does a `button->action = *action;` and a `button->action.macro = macro` if its a macro.

Line 1229 in `libratbag.c`

staticssleever668 replied:
Ah, literally the next line. Not thorough review, sorry...
Of course it's better to use the wrapper instead of directly modifying the underlying data.

staticssleever668 commented:
As for writing, the code seems OK, so I don't know what may be the reason.
Did you make and compare USB dumps made with original software and with this driver?

Daegalus commented:
> That's weird to say the least. How do you know they are read correctly?

I added `log_debug` messages all over the place. One in `read_profile` inside the button loop. Then in `read_button` i added a bunch inbetween the calls to button_to_action and such. Then I went even further adding the log_debugs to the loops in those functions. The numbers that showed up made sense based on the maps. 

Ill try to do some more debugging. I am seeing the logs by running `sudo ratbagd.devel --verbose=raw`

staticssleever668 commented:
Could you post the dumps made with official software and with libratbag? Won't hurt to have more eyes comparing them.
Comparing the changes made against the original Kone EMP driver - the button reading/writing procedure seems to be the same, only the values for actions are different. I doubt this is a broken feature on the Kone EMP driver, so maybe there actually is some other difference.

staticssleever668 commented on src/driver-roccat-kone-aimo.c:
```diff
@@ -0,0 +1,1220 @@
+					"Error while waiting for the device to be ready: %s (%d)\n",
+					strerror(-rc), rc);
+
+			bank_buf[0] = ROCCAT_REPORT_ID_MACRO;
+			bank_buf[1] = ROCCAT_BANK_ID_2;
+			// The remaining macro structure is not big enough to fill the second bank
+			// Write the remaining, fill the end with 0
+			unsigned int remaining_to_write = sizeof(struct roccat_macro)-ROCCAT_REPORT_SIZE_MACRO_BANK;
+			memcpy(bank_buf+2, &((uint8_t*)macro)[ROCCAT_REPORT_SIZE_MACRO_BANK], remaining_to_write);
+			memset(bank_buf+2+remaining_to_write, 0, ROCCAT_REPORT_SIZE_MACRO_BANK-(2+remaining_to_write));
```
```suggestion
			memset(bank_buf, 0, ROCCAT_REPORT_SIZE_MACRO_BANK);
			bank_buf[0] = ROCCAT_REPORT_ID_MACRO;
			bank_buf[1] = ROCCAT_BANK_ID_2;
			unsigned int remaining_to_write = sizeof(struct roccat_macro)-ROCCAT_REPORT_SIZE_MACRO_BANK;
			memcpy(bank_buf+2, &((uint8_t*)macro)[ROCCAT_REPORT_SIZE_MACRO_BANK], remaining_to_write);
```
Pointer arithmetic is dangerous and has to be done with great care - let's prevent it when possible.

Also, the `sizeof(struct roccat_macro)-ROCCAT_REPORT_SIZE_MACRO_BANK` and `bank_buf+2` look dodgy, too. I would like to see USB dumps, maybe I could figure something out.

staticssleever668 replied:
I may be misunderstanding something here, but if it works, replace `&((uint8_t*)macro)[ROCCAT_REPORT_SIZE_MACRO_BANK]` with `memcpy(bank_buf+2, (uint8_t*)(macro + ROCCAT_REPORT_SIZE_MACRO_BANK), remaining_to_write);` - makes the intent a bit clearer.

Daegalus commented:
Ok, I dumped everything into the files below. The files without extensions are the hex dumps, they should just be binary files, any binary file editor should work with them.

I use [ImHex](https://imhex.werwolv.net) (opensource, crossplatform), which lets me create pattern files that highlight and breakout the data for me to easily read it. All the `.pat` files can be imported into ImHex when the file is open and run against the file to highlight it and organize it.

Finally, the `roccat_dumps_all.txt` file has the roccat debug output for the 3 reports.

I was thinking that I didn't like all the pointer nonsense also, so I was wondering if maybe I copy over the structs I was using originally, using unions to combine them, and adjust the Macro code. I now realize how easy it is to get both banks. When I was wiresharking the usb packets, both banks got returned every time (though i realize these are writes). Now I can request each individual bank, and get the information as needed, making parsing it super easy using the old structs I had. Let me know if that would be better than this current 1 struct mangling situation. We also don't lose data from the 2nd packet this way.

[dumps_and_patterns.zip](https://github.com/libratbag/libratbag/files/9993868/dumps_and_patterns.zip)

![Screenshot from 2022-11-11 17-51-55](https://user-images.githubusercontent.com/670212/201452040-4cb327ea-d8a1-4dfb-b8b1-0ee1145ce934.png)
![Screenshot from 2022-11-11 17-48-45](https://user-images.githubusercontent.com/670212/201452041-7bee3c49-9fec-4158-8ca9-ac1b361f4d9d.png)
![Screenshot from 2022-11-11 17-45-16](https://user-images.githubusercontent.com/670212/201452042-a7a8ea6e-8a97-48bd-8efe-ffb3e87c30de.png)

staticssleever668 commented on src/driver-roccat-kone-aimo.c:
```diff
@@ -0,0 +1,1220 @@
+	uint8_t repeats;                            // Number of repetition for this macro
+	char group[ROCCAT_MACRO_GROUP_NAME_LENGTH]; // Folder name
+	char name[ROCCAT_MACRO_NAME_LENGTH];
+	uint16_t length;
```
Note: there is no `length` field in your `macro_hex_dump1.pat`.

Daegalus replied:
It's supposed to be there. I might've forgotten to save the pattern after fixing it. Without it, all the data is shifted incorrectly.

staticssleever668 commented:
> Let me know if that would be better than this current 1 struct mangling situation.

Sounds good.
By the way, looks like there may be no reason to have `macros` field in `struct roccat_data` as it's not accessed directly anywhere. When data is read from the mouse, it's put in this field, but parsing it just gets forgotten, and before writing it's even being filled with zeroes. So you can make them local variables in functions where it is used.
